hazelcast/src/main/java/com/hazelcast/spi/properties/GroupProperty.java

```diff
@@ -70,19 +73,40 @@
      * <p>
      * If this is less than the number of partitions on a member, partition operations
      * will queue behind other operations of different partitions.
-     * <p>
-     * The default is -1, which means that the value is determined dynamically.
-     */
+     *
+     * If the JVM detects 20 or more cores, it will prevent creating more threads than available cores by
+     * determining the number of operation threads as corecount-iothreadcount.
+     * This will prevent that IO threads and partition threads will contend for cores and this leads to
+     * increases throughput and less jitter.
+      */
     public static final HazelcastProperty PARTITION_OPERATION_THREAD_COUNT
-            = new HazelcastProperty("hazelcast.operation.thread.count", -1);
+            = new HazelcastProperty("hazelcast.operation.thread.count", new Function<HazelcastProperties, Integer>() {
+        @Override
+        public Integer apply(HazelcastProperties properties) {
+            int ioThreads = properties.getInteger(IO_INPUT_THREAD_COUNT) + properties.getInteger(IO_OUTPUT_THREAD_COUNT);
+            int availableProcessors = RuntimeAvailableProcessors.get();
+
+            if (availableProcessors >= 20 && ioThreads <= 8) {
+                // there are sufficient cores, lets isolate the IO threads from the partition threads.
+                return availableProcessors - ioThreads;
```

Comment:
so it machine has 18 cores, there will be 18 partition threads but if it has 20 cores then it will have 12 partition threads. this sounds unintuitive to me. 
Isn't having an explicit dedicated IO threads property better alternative, instead of this heuristic?

Reply:
If there are few cores, having insufficient number of partition threads can become (more of a) a utilization problem.

The premise of this change is that it should work out of the box. It should be enabled by default. What do you suggest?

Reply:
I am not a performance expert, how this affects querying (with or without offheap), aggregations, hot-restart etc. ? 
If do you think this should be enabled by default, I'm ok with that. But I think users should be able to disable it with a single switch too.

Reply:
In case of number crunching (so way more work done by partition threads compared to IO threads) this will have a negative performance impact because you will loose some partition threads and hence cores.

Reply:
I can add switch 'dedicated IO threads for big bix'. If set to default, it will try to be smart and show the current behavior.

If disabled explicitly, then it will return to HZ 3.x thread configuration.